Anchoring projections on recent data

nrmogh opened the pull request:
Let me know your thoughts about the input/output we should have here. At the moment it takes the calibration output and returns the subset of results and parameters that are within top x% in terms of difference from the data on a specified date. I thought this would be convenient because any further processing that happens to the projection results should be compatible with these outputs.

nrmogh committed:
Function for post processing projection trajectories based on final data point

nrmogh committed:
Function for post processing projection trajectories based on final data point

rlewinter commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+    runner_output: CalibrationOutput,
+    surveillance_data: pd.DataFrame,
+    top_fraction: float,
+    final_fitting_date: str | pd.Timestamp | datetime,
```
maybe "anchor_date"?

mu373 replied:
Is there any possibility that we also want to set the anchor to be a date range, not a single date? We could set the aguments as `anchor_start_date` and `anchor_end_date` so that it is general enough for such usages. For "point anchoring", the user can set the same date for the start and the end date.

rlewinter commented:
Recommendations

schema.calibration:
- new class `AnchoringSpec` + new `CalibrationConfiguration` field `anchoring`
- reuse info from `ComparisonSpec` for surveillance location/target/date and simulation target
- validate `top_fraction` and `distance_function`, ensure projection enabled

schema.general
- validate `final_fitting_date` against simulation dates

schema.dispatcher
- new class `AnchoringArguments` + new `BuilderOutput` field `anchoring`

dispatcher.builder.build_calibration
- construct `AnchoringArguments` from configs and add to `BuilderOutput`

dispatcher.runner.run_calibration_with_projection
- apply filtering, if exception occurs log warning and return results without filtering

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+    If `simulation_target` is a list, sums the simulation values at the fitting date
+    for all specified targets.
+
+    Args:
```
Needs conversion of docstring to numpy style

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+    filtered_projection_parameters = runner_output.results.projection_parameters["baseline"].iloc[idx]
+
+    # Summary statement
+    print(f"Filtered projections: {len(filtered_projections)} out of {len(runner_output.results.projections['baseline'])} trajectories remain after applying top_fraction={top_fraction}")
```
Use logger instead of direct print

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+        # Check all targets exist in trajectory
+        for target in simulation_target_list:
+            if target not in trajectory:
+                raise KeyError(f"Simulation trajectory missing expected target column '{target}'")
```
This check could be extracted outside the trajectory for-loop for optimization. As long as the trajectory is valid, I think checking for single trajectory is sufficient.

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+
+    # Compute distances
+    distances = []
+    for trajectory in runner_output.results.projections["baseline"]:
```
Not high but could be possible to have failed projections. `filter_failed_projections` could be used here.

https://github.com/mobs-lab/epymodelingsuite/blob/a25193026aa51179f97710130c087b2c3ae13fdb/epymodelingsuite/dispatcher/output.py#L35

mu373 replied:
We could also add checks that we have ≥1 valid projections after this filtering.

rlewinter replied:
On target-prop-ed I have applied filter_failed_projections() near the beginning of generate_calibration_outputs(), so this will cover all uses of projection results. Previously it had been called within loops over the calibration results for both quantiles and trajectories.

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+
+    # Get location and filter data
+    location = convert_location_name_format(runner_output.population, "ISO")
+    surveillance_location = get_data_in_location(
```
Could have better name for clarity e.g. `surveillance_data_at_location`. `surveillance_location` sounds like it contains the name of location.

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+from epymodelingsuite.utils.location import convert_location_name_format
+from .schema.dispatcher import CalibrationOutput
+from datetime import datetime
+
```
Needs importing of CalibrationOutput from schema for type definition

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,120 @@
+                raise KeyError(f"Simulation trajectory missing expected target column '{target}'")
+
+        try:
+            date_index = trajectory["date"].index(final_fitting_date)
```
Not 100% sure, but I think the trajectory always has same length for projections. In such case, `date_index` calculation can also be extracted outside the trajectory for-loop for optimization.

nrmogh committed:
Update filtering.py

nrmogh committed:
Update filtering.py

nrmogh committed:
Update calibration.py

nrmogh committed:
Update calibration.py

nrmogh committed:
Update dispatcher.py

nrmogh committed:
Update dispatcher.py

nrmogh committed:
Update general.py

nrmogh committed:
Update general.py

nrmogh committed:
Update builder.py

nrmogh committed:
Update builder.py

nrmogh committed:
Update runner.py

nrmogh committed:
Update runner.py

nrmogh commented:
Thanks for your the suggestions. Here's a summary of the changes I made. Let me know what you think. One thing I'm not sure about is if we might want to return both results (before and after anchoring). At the moment I just update the calibration results with the trajectories and parameters remaining after anchoring.
1. `filtering.py`
- Replaced final_fitting_date with anchor_start_date and anchor_end_date: The function now accepts a date range and compares vectors of all data points within that range. For point anchoring, users can set the same date for both parameters.
- Updated docstring to NumPy style
- Added logger rather than print
- Extracted target checks outside the loop
2. `schema.calibration.py`
- Created AnchoringSpec class with:
     - anchor_start_date and anchor_end_date for the date range
     - top_fraction validation (must be in (0, 1])
     - distance_function validation (must be one of: rmse, wmape, ae, mae, mape)
     - Optional fields that fall back to ComparisonSpec values if not provided
     - Added anchoring field to CalibrationConfiguration
     - Added validator to ensure anchoring requires projection to be enabled
3. `schema.general.py`
- Added validation in validate_cross_config_consistency to ensure anchor dates are within the projection end date range
4. `schema.dispatcher.py`
- Created AnchoringArguments class with all necessary fields for filtering
- Added anchoring field to BuilderOutput
5. `dispatcher.builder.py`
- Updated build_calibration to construct AnchoringArguments from config
- Falls back to first ComparisonSpec values when anchoring fields are not specified
- Added anchoring to BuilderOutput construction
6. `dispatcher.runner.py`
- Updated run_calibration_with_projection to apply anchoring after projection when anchoring is specified
- If filtering fails, logs a warning and returns unfiltered results

mu373 commented on epymodelingsuite/schema/calibration.py:
```diff
@@ -96,6 +96,41 @@ class ComparisonSpec(BaseModel):
+    anchor_start_date: date = Field(description="Start date (inclusive) of the anchor date range for filtering observed data")
+    anchor_end_date: date = Field(description="End date (inclusive) of the anchor date range for filtering observed data")
+    top_fraction: float = Field(description="Fraction of projections to keep (between 0 and 1)")
+    distance_function: str = Field(description="Distance function name for comparing data (e.g., 'rmse', 'mae', 'wmape')")
```
Let's extract distance_function as a separate Enum class, because now it is used in more than two places. (Other one in CalibrationConfiguration)

DistanceFunctionEnum:
- rmse
- mae
- wmape

mu373 commented on epymodelingsuite/schema/calibration.py:
```diff
@@ -96,6 +96,41 @@ class ComparisonSpec(BaseModel):
+
+    anchor_start_date: date = Field(description="Start date (inclusive) of the anchor date range for filtering observed data")
+    anchor_end_date: date = Field(description="End date (inclusive) of the anchor date range for filtering observed data")
+    top_fraction: float = Field(description="Fraction of projections to keep (between 0 and 1)")
```
I think top_fraction field needs to be optional.

mu373 commented on epymodelingsuite/schema/calibration.py:
```diff
@@ -96,6 +96,41 @@ class ComparisonSpec(BaseModel):
+    observed_location_column: str | None = Field(
+        None, description="Name of column containing location identifiers. If None, uses first comparison spec."
+    )
+    simulation: list[str] | None = Field(
```
`observed_*` and `simulation` are duplicates of `ComparisonSpec`, so I think we can reuse the class definition and have it under `comparison` field just as in `CalibrationConfiguration`.

```py
 class AnchoringSpec(BaseModel):
      """Specification for anchoring/filtering projections based on surveillance data."""

      anchor_start_date: date = Field(...)
      anchor_end_date: date = Field(...)
      top_fraction: float = Field(...)
      distance_function: str = Field(...)
      comparison: ComparisonSpec | None = Field(
          None, description="Comparison spec for anchoring. If None, uses first comparison spec from calibration."
      )
```

rlewinter replied:
For ComparisonSpec, are we not also using references to `options.surveillance`? Might need refactor

mu373 commented on epymodelingsuite/schema/dispatcher.py:
```diff
@@ -37,6 +37,18 @@ class ProjectionArguments(BaseModel):
         default=None, description="SMC generation number from which to draw parameter sets for projection."
     )
 
+class AnchoringArguments(BaseModel):
```
Having `Arguments` that corresponds to `Specs` aligns with `Projection`, but I don't remember why we had to define these separately from `Specs`. Do you remember if we had any reasons for this? @nrmogh @rlewinter

rlewinter replied:
At least for classes I wrote, I named them `Arguments` if their fields corresponded 1-to-1 with the arguments of a function, so we are using the class to provide the arguments for a single function call. I used `Specs` if the class contained more general information about a task.

rlewinter replied:
It looks like AnchoringSpec and AnchoringArguments are exactly the same, so there's no need for both of them.

mu373 commented on epymodelingsuite/filtering.py:
```diff
@@ -0,0 +1,183 @@
+    for idx, date_val in enumerate(trajectory_dates):
+        date_ts = pd.Timestamp(date_val)
+        if anchor_start_date <= date_ts <= anchor_end_date:
+            date_indices.append(idx)
```
Date indices extraction can be optimized without  using for-loop.
```py
trajectory_dates = pd.to_datetime(first_trajectory["date"])
date_mask = (anchor_start_date <= trajectory_dates) & (trajectory_dates <= anchor_end_date)
date_indices = np.where(date_mask)[0].tolist()
```

mu373 commented on epymodelingsuite/dispatcher/runner.py:
```diff
@@ -4,12 +4,23 @@
 logger = logging.getLogger(__name__)
 
+# Distance function dictionary matching builder
+dist_func_dict = {
```
This can be imported from dispatcher/builder, rather than defining duplicate.
```py
# Remove importing of distance functions
# from epydemix.calibration import ae, mae, mape, rmse, wmape
from .runner import dist_func_dict
```

mu373 commented on epymodelingsuite/dispatcher/runner.py:
```diff
@@ -210,6 +221,56 @@ def run_calibration_with_projection(
+                
+                # Apply filtering
+                original_count = len(projection_results.projections["baseline"])
+                filtered_projections, filtered_projection_parameters = filter_projections_on_data_point(
```
The latest function name defined in filtering.py is `anchor_projections_on_data()`. Also needs import update.

mu373 commented on epymodelingsuite/dispatcher/runner.py:
```diff
@@ -210,6 +221,56 @@ def run_calibration_with_projection(
+                    simulation_target=configs.anchoring.simulation_target,
+                )
+                
+                # Update projection results with filtered data
+                projection_results.projections["baseline"] = filtered_projections
+                projection_results.projection_parameters["baseline"] = filtered_projection_parameters
```
I would vote for saving filtered results under separate field from `results` in `CalibrationOutput`. Overriding the original results with filtered data removes the full information that we originally had, meaning that it limits the possibility for additional post-processes (e.g. anchoring with different distance functions, top fraction percentiles, etc.).

mu373 reviewed (commented):
Great work! Small changes are as commented in-line. As I've commented below, I would vote for saving filtered results under separate field from results in CalibrationOutput, since overriding is a lossy/irreversible operation.

I haven't ran the codes written here yet, but it would be great to have the tutorial filled in so that we can confirm that it works without raising errors.

rlewinter reviewed (commented):
Along with returning both original and anchored projections, we need to specify in output config which are used to create outputs. I think it's fine to only use one per run of output generator.

Not sure about this one, but there might be some relevant info we'd want to put in the metadata file.

mu373 commented:
> we need to specify in output config which are used to create outputs

One of the ideas I have in mind is to have the definition of anchoring be made in output config, not in the modelset. I think it makes sense to consider anchoring as part of "post-processing" and to run it inside the output stage, while the runner just focuses on the main calibration/projection.

rlewinter commented:
> One of the ideas I have in mind is to have the definition of anchoring be made in output config, not in the modelset. I think it makes sense to consider anchoring as part of "post-processing" and to run it inside the output stage, while the runner just focuses on the main calibration/projection.

Agreed

nrmogh committed:
plotting notebooks

nrmogh committed:
plotting notebooks

nrmogh committed:
added handling for covid vaccine data

nrmogh committed:
added handling for covid vaccine data

nrmogh committed:
Allow sampling omega and eta in months

nrmogh committed:
Allow sampling omega and eta in months

nrmogh committed:
save projection sampled params

nrmogh committed:
save projection sampled params

nrmogh committed:
updates

nrmogh committed:
Merge branch 'nima-dev' of https://github.com/mobs-lab/epymodelingsuite into nima-dev

nrmogh committed:
Support np.exp in calculated beta

nrmogh committed:
stop recording projection parameters in trajectories. this is now handled seperately

nrmogh committed:
remove hardcoded month parameters

nrmogh committed:
new tests

nrmogh committed:
Climate seasonality

nrmogh committed:
rescale climate seasonality to s_min - 1.0

nrmogh committed:
add mobility to seasonality function

nrmogh committed:
reproducible projections

nrmogh committed:
Update core.py

nrmogh committed:
flexible xlimits for plots

nrmogh committed:
Update core.py

nrmogh committed:
plotting sum of multiple transitions
